src/cdk/table/table.ts

```diff
@@ -324,6 +324,18 @@ export class CdkTable<T> implements AfterContentChecked, CollectionViewer, OnDes
    */
   private _cachedRenderRowsMap = new Map<T, WeakMap<CdkRowDef<T>, RenderRow<T>[]>>();
 
+  /**
+   * Cache of the current rendering rows by render index and their row definition (CdkRowDef)
+   * that was used to render them.
+   *
+   * This will update every time the differ emits a change and will support detection of a change
+   * in the row definition without a change in the context or identity which did not result
+   * in a differ operation. This is most common when the `trackBy` method return the `index` which
+   * will prevent the differ from adding/removing/moving rows so a change in the rowDef
+   * will be missed.
+   */
+  private _cachedRenderDefMap = new Map<number, CdkRowDef<T>>();
```

Comment:
```suggestion
  private _cachedRenderDefMap = new WeakMap<number, CdkRowDef<T>>();
```
It looks like this can be a `WeapMap` since we never have to enumerate it

Reply:
I don't think that primitives are allowed in WeakMap.
After all, WeakMap is based on references and number does not have a reference and the GC will not be able to actually do anything with it...

Reply:
Good point, I wasn't thinking about that.